recipes/cuda-cudart/meta.yaml

```
@@ -0,0 +1,290 @@
+{% set name = "cuda-cudart" %}
+{% set version = "12.0.107" %}
+{% set cuda_version = "12.0" %}
+{% set platform = "linux-x86_64" %}  # [linux64]
+{% set platform = "linux-ppc64le" %}  # [ppc64le]
+{% set platform = "linux-sbsa" %}  # [aarch64]
+{% set platform = "windows-x86_64" %}  # [win]
+{% set target_name = "x86_64-linux" %}  # [linux64]
+{% set target_name = "ppc64le-linux" %}  # [ppc64le]
+{% set target_name = "sbsa-linux" %}  # [aarch64]
+{% set target_name = "x64" %}  # [win]
+{% set extension = "tar.xz" %}  # [not win]
+{% set extension = "zip" %}  # [win]
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://developer.download.nvidia.com/compute/cuda/redist/cuda_cudart/{{ platform }}/cuda_cudart-{{ platform }}-{{ version }}-archive.{{ extension }}
+  sha256: c0e16e9593a69afe4997ae96a8b024e751ebce0254523fd969b08bb028582d23  # [linux64]
+  sha256: d785fd613e35e3761fdb0016a31a56417f570605cbfb9a93ca9dc68be6789188  # [ppc64le]
+  sha256: 5de248c6c203f6c91ce34256bbfc028bc87c1b0c552f9f51edb65e60ad665f63  # [aarch64]
+  sha256: 1cd8516feee068c6d718453e76b0dfcbee27844a35d4f3aa608a3d316190309c  # [win]
+
+build:
+  number: 0
+  skip: true  # [osx]
```

Windows CI appears to be failing because of overlinking.

```
conda_build.exceptions.OverLinkingError: overlinking check failed 
['  ERROR (cuda-cudart,Library/bin/cudart64_12.dll): $RPATH/api-ms-win-core-libraryloader-l1-2-0.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  ERROR (cuda-cudart,Library/bin/cudart64_12.dll): $RPATH/api-ms-win-security-systemfunctions-l1-1-0.dll not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?']
##[error]Cmd.exe exited with code '1'.
```

Seems related to the problems mentioned in previous PRs here:
- https://github.com/conda-forge/staged-recipes/pull/21924#issuecomment-1480427413
- https://github.com/conda-forge/staged-recipes/pull/21924#issuecomment-1480449462
- https://github.com/conda-forge/staged-recipes/pull/21913

The solution in #21924 was disabling the overlinking check. Is this what we should do, or is there a different approach?
```suggestion
build:
  number: 0
  skip: true  # [osx]
  error_overlinking: false  # [win]
```

---

At one point the recipes in the aforementioned PR had `error_overlinking` ( https://github.com/conda-forge/staged-recipes/pull/21924/commits/3ed95cfb099c3474fd0ebc0d63e405090a03cb01 ), but it seems to be reverted later ( https://github.com/conda-forge/staged-recipes/pull/21924/commits/96f6ad09ca0b60fe54c6143139eb14e68c55bae2 ). Am not seeing `error_overlinking` in the final changes in that PR or in the recipes. So think Leo came up with another solution, but it is not obvious what it was based on the commit history of the PR what it was. Might be worth studying the `libnvjitlink` recipe more closely (since this is where it was needed and later dropped)

---

Hmm. I looked over the [diff between those commits](https://github.com/conda-forge/staged-recipes/compare/3ed95cfb099c3474fd0ebc0d63e405090a03cb01..96f6ad09ca0b60fe54c6143139eb14e68c55bae2) and compared the requirements of those packages to `cuda-cudart` but I'm not seeing much that points to the original cause or solution.

---

Yeah it wasn't obvious to me how this was fixed after a cursory look. Though assumed I had just missed something

Let's try merging in `main` (as we discussed offline) and see if the issue persists

---

Still seeing the same issue after updating the branch. This [Microsoft support thread]( https://social.msdn.microsoft.com/Forums/en-US/a28331ae-19a3-4a34-b3ba-1e8fd4430375/missing-apimswincore-dlls ) looks relevant

---

So I do see `api-ms-win-core-libraryloader-*.dll` in [the `vc` package]( https://github.com/conda-forge/vc-feedstock/blob/cb1cb5e46107423c469ff50ca6a8e7187051f6a0/recipe/meta.yaml#L78 ), but the SONAME(?) appears different. Maybe we are getting the wrong version?

With `api-ms-win-security-systemfunctions-*.dll`, it doesn't not appear to be in that list. Unclear as to whether it should be (possibly in a newer version?)

---

My impression was that after I followed the libcublas approach and added compiler to build, I no longer saw warnings on the dlls, so I assumed it's safe to undo the WAR and gave it a shot. 

Another issue was that I noticed the internal recipe had a wrong cusparse/nvjitlink version, mismatching the CUDA 12.0 one, so I fixed it and that change might also explain. Not sure which is the real key.

---

Thanks Leo! 🙏 Really appreciate you taking time to chime in here 🙂

Yeah the `compiler`s part seems like the key. Did look at that. Though this recipe already seems to include these for multiple packages.

That said, I do see `noarch: generic` in a few places where we are adding `compiler`s and doing Linux/Windows specific things. Maybe these are holdovers from before we started doing architecture specific things. Not sure if they are causing this problem, but would expect them to cause some issues. So we may want to drop them. Have outlined them in review comments below

---

Discussed this overlinking issue with @jakirkham / @adibbley. We agreed to skip Windows for now, and then work on fixing this problem in the feedstock. That way Linux packages can be unblocked and we can continue on `cuda-nvcc` while we investigate the Windows overlinking further.
```suggestion
build:
  number: 0
  skip: true  # [osx or win]
```

---

Raised feedstock issue ( https://github.com/conda-forge/cuda-cudart-feedstock/issues/1 ) for further discussion/investigation.